presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -0,0 +1,481 @@
+                }
+                else {
+                    throw new IllegalArgumentException("Unrecognized memory pool: " + memoryPool);
+                }
+            }
+        }
+
+        ImmutableMap.Builder<MemoryPoolId, ClusterMemoryPoolInfo> memoryPoolInfos = ImmutableMap.builder();
+        ClusterMemoryPool pool = new ClusterMemoryPool(GENERAL_POOL);
+        pool.update(memoryInfos, queriesAssignedToGeneralPool);
+        ClusterMemoryPoolInfo clusterInfo = pool.getClusterInfo(Optional.ofNullable(largestGeneralPoolQuery).map(Query::getQueryId));
+        memoryPoolInfos.put(GENERAL_POOL, clusterInfo);
+        if (isReservedPoolEnabled) {
+            pool = new ClusterMemoryPool(RESERVED_POOL);
+            pool.update(memoryInfos, queriesAssignedToReservedPool);
+            memoryPoolInfos.put(RESERVED_POOL, pool.getClusterInfo());
+        }
+        return memoryPoolInfos.build();
+    }
+
+    private Query getLargestMemoryQuery(Optional<Query> existingLargeQuery, Query newQuery)
+    {
+        requireNonNull(newQuery, "newQuery must not be null");
+        return existingLargeQuery
+                .map(largeQuery -> {
+                    long largestGeneralBytes = largeQuery.getBasicQueryInfo().getQueryStats().getTotalMemoryReservation().toBytes();
+                    long currentGeneralBytes = newQuery.getBasicQueryInfo().getQueryStats().getTotalMemoryReservation().toBytes();
+                    if (currentGeneralBytes > largestGeneralBytes) {
+                        return newQuery;
+                    }
+                    return largeQuery;
+                })
+                .orElse(newQuery);
+    }
+
+    public Map<String, MemoryInfo> getWorkerMemoryInfo()
+    {
+        return nodeStatuses.entrySet().stream().collect(toImmutableMap(e -> {
+            String nodeIdentifier = e.getValue().getNodeStatus().getNodeId();
+            String nodeHost = URI.create(e.getValue().getNodeStatus().getExternalAddress()).getHost();
+            return nodeIdentifier + " [" + nodeHost + "]";
+        }, e -> e.getValue().getNodeStatus().getMemoryInfo()));
+    }
+
+    private void validateCoordinatorConsistency()
+    {
+        Set<String> coordinators = internalNodeManager.getCoordinators().stream().map(InternalNode::getNodeIdentifier).collect(toImmutableSet());
+        Set<String> heartbeatedCoordinatorNodes = nodeStatuses.values().stream()
+                .map(InternalNodeState::getNodeStatus)
+                .filter(NodeStatus::isCoordinator)
+                .map(NodeStatus::getNodeId)
+                .collect(toImmutableSet());
+        if (!(Sets.difference(coordinators, heartbeatedCoordinatorNodes).isEmpty() && !coordinators.isEmpty())) {
+            throw new ResourceManagerInconsistentException(format("%s nodes found in discovery vs. %s nodes found in heartbeats", coordinators.size(), heartbeatedCoordinatorNodes.size()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;size&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 297. Consider formatting this data with &#39;%d&#39; instead.